src/agent_observer.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent relation observer module."""
+
+
+import logging
+import os
+from pathlib import Path
+
+import ops
+
+import service
+from charm_state import AGENT_RELATION, State
+
+logger = logging.getLogger()
+JENKINS_WORKDIR = Path("/var/snap/jenkins-agent/common/agent")
+AGENT_READY_PATH = Path(JENKINS_WORKDIR / ".ready")
+
+
+class Observer(ops.Object):
+    """The Jenkins agent relation observer."""
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        state: State,
+        jenkins_agent_service: service.JenkinsAgentService,
+    ):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+            jenkins_agent_service: Service manager that controls Jenkins agent service.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+        self.jenkins_agent_service = jenkins_agent_service
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_joined, self._on_agent_relation_joined
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_changed, self._on_agent_relation_changed
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
+        )
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired when an agent has joined the relation.
+        """
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.debug("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
```

Comment:
I'm not following this, when the relation is joined we take the data from the relation and update it into the event? I was guessing we would start an agent service?

Reply:
I believe the way the integration is currently implemented requires a 2-way interaction, meaning that both jenkins and the agent charm need to set values on their side of the relation databag. I documented this interaction and request that we change it in this ticket: https://warthogs.atlassian.net/browse/ISD-1404

Reply:
Ah I see, perhaps worth a comment/ explanation around this. I do remember that yeh

Reply:
That may influence/ explain why in other parts of the code we are ok with incomplete relation data